### PR TITLE
CASSANDRA-16959 trunk: Immediatelly apply stream throughput, considering negative values

### PR DESCRIPTION
**Pull request by adelapena**

### CODE REVIEW

File: src/java/org/apache/cassandra/streaming/StreamManager.java [hunk 1]
```diff
@@ -62,41 +63,62 @@ public static StreamRateLimiter getRateLimiter(InetAddressAndPort peer)
...
+        private static double calculateRateInBytes()
+        {
+            return DatabaseDescriptor.getStreamThroughputOutboundMegabitsPerSec() * BYTES_PER_MEGABIT > 0
+                   ? DatabaseDescriptor.getStreamThroughputOutboundMegabitsPerSec() * BYTES_PER_MEGABIT
+                   : Double.MAX_VALUE; // if throughput is set to 0, throttling is disabled
+        }
+
+        private static double calculateInterDCRateInBytes()
+        {
+            return DatabaseDescriptor.getInterDCStreamThroughputOutboundMegabitsPerSec() * BYTES_PER_MEGABIT > 0
```
Comment (bereng) [inline src/java/org/apache/cassandra/streaming/StreamManager.java]: `* BYTES_PER_MEGABIT` will never be zero or negative so I guess you could simplify the conditional for readability and sparing an operation?

`return DatabaseDescriptor.getInterDCStreamThroughputOutboundMegabitsPerSec() * BYTES_PER_MEGABIT > 0 ?`
->
`return DatabaseDescriptor.getInterDCStreamThroughputOutboundMegabitsPerSec() > 0 ?`

File: src/java/org/apache/cassandra/streaming/StreamManager.java [hunk 2]
```diff
@@ -62,41 +63,62 @@ public static StreamRateLimiter getRateLimiter(InetAddressAndPort peer)
...
         public void acquire(int toTransfer)
         {
             limiter.acquire(toTransfer);
             if (!isLocalDC)
                 interDCLimiter.acquire(toTransfer);
         }
+
+        public static synchronized void updateThroughput()
+        {
+            limiter.setRate(calculateRateInBytes());
```
Comment (azotcsit) [inline src/java/org/apache/cassandra/streaming/StreamManager.java:89]: `setRate` seems to be thread-safe (see https://github.com/google/guava/blob/v27.0/guava/src/com/google/common/util/concurrent/RateLimiter.java#L227), so I'm not sure we need `synchronized` here. WDYT?

File: src/java/org/apache/cassandra/streaming/StreamManager.java [hunk 3]
```diff
@@ -62,41 +63,62 @@ public static StreamRateLimiter getRateLimiter(InetAddressAndPort peer)
...
+        public static synchronized void updateInterDCThroughput()
+        {
+            interDCLimiter.setRate(calculateInterDCRateInBytes());
+        }
+
+        private static double calculateRateInBytes()
+        {
+            return DatabaseDescriptor.getStreamThroughputOutboundMegabitsPerSec() * BYTES_PER_MEGABIT > 0
+                   ? DatabaseDescriptor.getStreamThroughputOutboundMegabitsPerSec() * BYTES_PER_MEGABIT
+                   : Double.MAX_VALUE; // if throughput is set to 0, throttling is disabled
```
Comment (azotcsit) [inline src/java/org/apache/cassandra/streaming/StreamManager.java]: nit: "if throughput is set to 0" -> "if throughput is set to 0 or negative value"

File: src/java/org/apache/cassandra/streaming/StreamManager.java (hunk 1 above)
Comment (adelapena) [reply src/java/org/apache/cassandra/streaming/StreamManager.java]: Good idea, done.

File: src/java/org/apache/cassandra/streaming/StreamManager.java (hunk 2 above)
Comment (adelapena) [reply src/java/org/apache/cassandra/streaming/StreamManager.java:89]: Right, we don't need synchronized here

**Pull request closed without merging**